### ISSUE TITLE
Unexpected Window title: FocusProxy

### ISSUE THREAD

**Issue by feluxe** [autokey/autokey#113]
## Classification:

Bug

## Reproducibility:

Always

## Summary

I use an IDE called *JetBrains PyCharm*. It produces windows with titles such as this one:

    MyProjectName - PyCharm

To confirm this I ran `xprop` on the PyCharm window. This is from the `xprop` output:

    WM_CLASS(STRING) = "sun-awt-X11-XFramePeer", "jetbrains-pycharm"
    WM_CLIENT_LEADER(WINDOW): window id # 0x6c00008
    _NET_WM_ICON_NAME(UTF8_STRING) = "MyProjectName  - PyCharm"
    WM_ICON_NAME(STRING) = "MyProjectName  - PyCharm"
    _NET_WM_NAME(UTF8_STRING) = "MyProjectName  - PyCharm"
    WM_NAME(STRING) = "MyProjectName  - PyCharm"

It looks fine so far, IMO.

Now if I use the `Display Window Info` entry in the icon context menu of *autokey*, I get this for the PyCharm window:

    Active window information:
    Title: 'FocusProxy'
    Class: 'Focus-Proxy-Window.FocusProxy'

This is a problem, because I cannot filter window titles with this strange output.

## Steps to Reproduce

Open PyCharm window. Run `Display Window Info` on PyCharm window.

## Expected Results

The window title should be `MyProjectName - PyCharm`

## Actual Results

The window title is `FocusProxy`

## Version

0.93.10

It didn't work with [autokey](https://github.com/autokey/autokey) either.

Installed via: pip

Distro: Tested with Linux Mint and Arch Linux.

**Comment by troxor** [comment]
Maybe related to #75 ? Isn't PyCharm a Java app?

**Comment by feluxe** [comment]
Yes it's a Java app. I'm wondering how does autokey get the window titles? If autokey uses a third party package to get the window titles, maybe I should file a bug there?

**Comment by troxor** [comment]
Autokey is based on python-xlib. What properties, if any, from `xprop` correspond to the information above?

**Comment by feluxe** [comment]
I just ran `xprop` in a terminal on my linux machine and then clicked on a PyCharm window with my mouse pointer. The PyChram window information appears in the terminal window after I click. This is the full output from xprop:
```
_MOTIF_DRAG_RECEIVER_INFO(_MOTIF_DRAG_RECEIVER_INFO) = 0x6c, 0x0, 0x5, 0x0, 0x92, 0x2f, 0xc0, 0x6, 0x0, 0x0, 0x0, 0x0, 0x10, 0x0, 0x0, 0x0
XdndAware(ATOM) = BITMAP
_NET_WM_DESKTOP(CARDINAL) = 0
WM_STATE(WM_STATE):
		window state: Normal
		icon window: 0x0
_NET_WM_VISIBLE_NAME(UTF8_STRING) = " my_project  - PyCharm"
_NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_NORMAL
_MOTIF_WM_HINTS(_MOTIF_WM_HINTS) = 0x3, 0x1, 0x1, 0x0, 0x0
_NET_WM_STATE(ATOM) = _NET_WM_STATE_MAXIMIZED_VERT, _NET_WM_STATE_MAXIMIZED_HORZ
_NET_FRAME_EXTENTS(CARDINAL) = 3, 3, 24, 3
WM_HINTS(WM_HINTS):
		Client accepts input or input focus: False
		Initial state is Normal State.
_NET_WM_ICON(CARDINAL) = 	Icon (32 x 32): (ICON REMOVED)
_NET_WM_PID(CARDINAL) = 18636
WM_CLIENT_MACHINE(STRING) = "peter-laptop"
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, WM_TAKE_FOCUS
WM_CLASS(STRING) = "sun-awt-X11-XFramePeer", "jetbrains-pycharm"
WM_CLIENT_LEADER(WINDOW): window id # 0x6c00008
_NET_WM_ICON_NAME(UTF8_STRING) = "my_project  - PyCharm"
WM_ICON_NAME(STRING) = "my_project  - PyCharm"
_NET_WM_NAME(UTF8_STRING) = "my_project  - PyCharm"
WM_NAME(STRING) = "my_project  - PyCharm"
WM_NORMAL_HINTS(WM_SIZE_HINTS):
		user specified location: -3, 27
		program specified location: -3, 27
		program specified size: 1362 by 714
		window gravity: NorthWest
```

This is the full output that I get from AutoKey's `Display Window Info` command:

```
Active window information:
Title: 'b'FocusProxy''
Class: 'Focus-Proxy-Window.FocusProxy'
```
I'm not sure if this answers your question. Please tell me if you want more information.

I'm not sure how this is related to #75 . I tested the `insert date` script and it works fine in the PyCharm window. I configured a bunch of shortcuts in autokey for PyCharm and those work fine as well (it's a godsend, actually). The only issue I have is that autokey uses unexpected window titles for PyCharm. My workaround is to just use `FocusProxy` as a window filter. I can live with that, since PyCharm is the only Java application that I use, so that the window filter is unique. If no one else is effected by this, I would probably leave the issue idle.

**Comment by troxor** [comment]
Perhaps related to this? https://bugs.java.com/bugdatabase/view_bug.do?bug_id=6961123 What version of JDK?

**Comment by feluxe** [comment]
I'm not sure if this is related, since I have not used java or python-xlib. But as long as autokey uses python-xlib the right way to get window title information, I assume that either PyCharm, python-xlib or JDK is responsible for the issue. 

```
JRE: 1.8.0_152-release-1024-b8 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Linux 4.14.8-1-ARCH
```
Since I get valid information from xprop, my guess is that `python-xlib` has issues with java apps.

On the other hand, if I run this ([snippet found here](https://unix.stackexchange.com/questions/179769/finding-certain-x-window-with-pythons-xlib)):

```python
from Xlib.display import Display

def printWindowHierrarchy(window, indent):
    children = window.query_tree().children
    for w in children:
        print(indent, window.window.get_wm_class())
        printWindowHierrarchy(w, indent+'-')

display = Display()
root = display.screen().root
printWindowHierrarchy(root, '-')
```

I see a bunch of lines like this one:

    --- ('sun-awt-X11-XFramePeer', 'jetbrains-pycharm')

And if I replace `window.get_wm_class` with `window.get_wm_name` in the snippet above, I actually get the right window titles from python-xlib:

    myproject - PyCharm

I have no idea where autokey gets `Title: 'FocusProxy'` from.

**Comment by luziferius** [comment]
I tracked it down a bit:
 https://github.com/autokey/autokey/blob/a12540a00557ed5e96d7efeb138dfdba7eecf916/lib/autokey/interface.py#L969-L972
This function receives a None window.
I inserted a `logger.info("get_window_title: window is None!")`right after that if and got this output. The logic then obtains the window properties of some ominous `localDisplay.get_input_focus().focus` which then yields the results you see.

**Comment by luziferius** [comment]
@feluxe This is fixed in this bugfix branch here https://github.com/autokey/autokey/tree/bugfix/issue_113, which will be included in the next release.

On my system, the log now states:
```
2018-12-07 23:13:31,547 DEBUG - service - Window visible title: 'autokey [~/Projekte/autokey] - .../lib/autokey/service.py [autokey] - PyCharm', Window class: 'sun-awt-X11-XFramePeer.jetbrains-pycharm-ce'
```
@watawuwu Thank you for your research. That helped me to write the fix for AutoKey.

**Comment by luziferius** [comment]
Released version [`0.95.5`](https://github.com/autokey/autokey/releases/tag/v0.95.5), containing the bug fix.

**Comment by vadimhp** [comment]
Running PHPStorm on Ubuntu 18.04, and displaying Active window information:
Title: 'FocusProxy'
Class: 'Focus-Proxy-Window.FocusProxy'
